### PR TITLE
Cherwell new integration

### PR DESCRIPTION
**Pull request by ShahafBenYakir**
## Status
Ready

## Related Issues
https://github.com/demisto/etc/issues/8109

## Description
Created integration for Cherwell Service Management 

## Does it break backward compatibility?
   - No

## Must have
- [x] Tests
- [x] Documentation - https://github.com/demisto/etc/issues/16735
- [ ] Code Review

### CODE REVIEW

File: Integrations/Cherwell/Cherwell.py [hunk 1]
```diff
@@ -0,0 +1,992 @@
...
+    if response.status_code == HTTP_CODES['unauthorized']:
+        token = get_access_token(True)
+        response = http_request(method, url, payload, token, custom_headers=headers)
+    return response
+
+
+def get_business_object_summary_by_name(name):
+    url = BASE_URL + f'api/V1/getbusinessobjectsummary/busobname/{name}'
+    response = make_request('GET', url)
+    res_json = parse_response(response, "Could not get business object summary")
```
Comment (yuvalbenshalom) [inline Integrations/Cherwell/Cherwell.py]: you can return directly

File: Integrations/Cherwell/Cherwell.py [hunk 2]
```diff
@@ -0,0 +1,992 @@
...
+    if not res:
+        return_error(f'Could not retrieve "{name}" business object id. '
+                     f'Make sure "{name}" is a valid business object.')
+    return res[0].get('busObId')
+
+
+def save_business_object(payload):
+    url = BASE_URL + "api/V1/savebusinessobject"
+    response = make_request("POST", url, json.dumps(payload))
+    res_json = parse_response(response, "Could not save business object")
```
Comment (yuvalbenshalom) [inline Integrations/Cherwell/Cherwell.py]: return directly

File: Integrations/Cherwell/Cherwell.py [hunk 3]
```diff
@@ -0,0 +1,992 @@
...
+    return res_json
+
+
+def get_business_object_record(business_object_id, object_id, id_type):
+    id_type_str = 'publicid' if id_type == 'public_id' else 'busobrecid'
+    url = BASE_URL + f'api/V1/getbusinessobject/busobid/{business_object_id}/{id_type_str}/{object_id}'
+    response = make_request("GET", url)
+    res_json = parse_response(response, "Could not get business objects")
```
Comment (yuvalbenshalom) [inline Integrations/Cherwell/Cherwell.py]: return directly

File: Integrations/Cherwell/Cherwell.py [hunk 4]
```diff
@@ -0,0 +1,992 @@
...
+    url = BASE_URL + f'api/V1/deletebusinessobject/busobid/{business_object_id}/{id_type_str}/{object_id}'
+    response = make_request("DELETE", url)
+    res_json = parse_response(response, "Could not delete business object")
+    return res_json
+
+
+def get_search_results(payload):
+    url = BASE_URL + "api/V1/getsearchresults"
+    response = make_request("POST", url, json.dumps(payload))
+    res_json = parse_response(response, "Could not search for business objects")
```
Comment (yuvalbenshalom) [inline Integrations/Cherwell/Cherwell.py]: return directly

File: Integrations/Cherwell/Cherwell.py [hunk 5]
```diff
@@ -0,0 +1,992 @@
...
+        "busObId": business_object_id,
+        "includeAll": include_all
+    }
+
+    if field_names:
+        payload['fieldNames'] = field_names
+    if fields_ids:
+        payload['fieldIds'] = fields_ids
+    response = make_request("POST", url, json.dumps(payload))
+    res_json = parse_response(response, "Could not get business object template")
```
Comment (yuvalbenshalom) [inline Integrations/Cherwell/Cherwell.py]: return directly

File: Integrations/Cherwell/Cherwell.py [hunk 6]
```diff
@@ -0,0 +1,992 @@
...
+    if object_id:
+        id_key = 'busObPublicId' if id_type == 'public_id' else 'busObRecId'
+        business_object_json[id_key] = object_id
+    return business_object_json
+
+
+def create_business_object(name, data_json):
+    business_object_id = resolve_business_object_id_by_name(name)
+    business_object_json = build_business_object_json(data_json, business_object_id)
+    result = save_business_object(business_object_json)
```
Comment (yuvalbenshalom) [inline Integrations/Cherwell/Cherwell.py]: return directly

File: Integrations/Cherwell/Cherwell.py [hunk 7]
```diff
@@ -0,0 +1,992 @@
...
+    return result
+
+
+def update_business_object(name, data_json, object_id, id_type):
+    business_object_id = resolve_business_object_id_by_name(name)
+    business_object_json = build_business_object_json(data_json, business_object_id, object_id, id_type)
+    result = save_business_object(business_object_json)
```
Comment (yuvalbenshalom) [inline Integrations/Cherwell/Cherwell.py]: return directly

File: Integrations/Cherwell/Cherwell.py [hunk 8]
```diff
@@ -0,0 +1,992 @@
...
+    results = get_business_object_record(business_object_id, object_id, id_type)
+    parsed_business_object = parse_fields_from_business_object(results.get('fields'))
+    parsed_business_object['PublicId'] = results.get('busObPublicId')
+    parsed_business_object['RecordId'] = results.get('busObRecId')
+    return parsed_business_object, results
+
+
+def delete_business_object(name, object_id, id_type):
+    business_object_id = resolve_business_object_id_by_name(name)
+    results = delete_business_object_record(business_object_id, object_id, id_type)
```
Comment (yuvalbenshalom) [inline Integrations/Cherwell/Cherwell.py]: return directly

File: Integrations/Cherwell/Cherwell.py [hunk 9]
```diff
@@ -0,0 +1,992 @@
...
+
+
+def download_attachment_from_business_object(attachment):
+    attachment_id = attachment.get('attachmentId')
+    business_object_id = attachment.get('busObId')
+    business_record_id = attachment.get('busObRecId')
+    url = BASE_URL + f'api/V1/getbusinessobjectattachment' \
+        f'/attachmentid/{attachment_id}/busobid/{business_object_id}/busobrecid/{business_record_id}'
+    response = make_request('GET', url)
+    attachment_content = parse_response(response, f'Unable to get content of attachment {attachment_id}')
```
Comment (yuvalbenshalom) [inline Integrations/Cherwell/Cherwell.py]: return directly

File: Integrations/Cherwell/Cherwell.py [hunk 10]
```diff
@@ -0,0 +1,992 @@
...
+    return get_search_results(payload)
+
+
+def get_key_value_dict_from_template(key, val, business_object_id):
+    template_dict = get_business_object_template(business_object_id)
+    business_object_ids_dict = cherwell_dict_parser(key, val, template_dict.get('fields'))
+    return business_object_ids_dict
+
+
+def get_all_incidents(objects_names, last_created_time, max_result, query_string):
```
Comment (yuvalbenshalom) [inline Integrations/Cherwell/Cherwell.py]: change to `max_results`

File: Integrations/Cherwell/Cherwell.py [hunk 11]
```diff
@@ -0,0 +1,992 @@
...
+
+def upload_business_object_attachment(file_name, file_size, file_content, object_type_name, id_type, object_id, ):
+    id_type_str = 'publicid' if id_type == 'public_id' else 'busobrecid'
+    url = BASE_URL + f'/api/V1/uploadbusinessobjectattachment/' \
+        f'filename/{file_name}/busobname/{object_type_name}/{id_type_str}/{object_id}/offset/0/totalsize/{file_size}'
+    payload = file_content
+    headers = HEADERS
+    headers['Content-Type'] = "application/octet-stream"
+    response = make_request('POST', url, payload, headers)
+    parsed_response = parse_response(response, f'Could not upload attachment {file_name}')
```
Comment (yuvalbenshalom) [inline Integrations/Cherwell/Cherwell.py]: return directly

File: Integrations/Cherwell/Cherwell.py [hunk 12]
```diff
@@ -0,0 +1,992 @@
...
+
+    validate_query_list(query_list)
+    return query_list
+
+
+def query_business_object_string(business_object_name, query_string, max_results):
+    if max_results:
+        try:
+            int(max_results)
+        except Exception:
```
Comment (yuvalbenshalom) [inline Integrations/Cherwell/Cherwell.py]: better:
`except ValueError:`

File: Integrations/Cherwell/Cherwell.py [hunk 13]
```diff
@@ -0,0 +1,992 @@
...
+    return business_object_info
+
+
+########################################################################################################################
+'''
+Commands
+'''
+
+
+def test_command():
```
Comment (yuvalbenshalom) [inline Integrations/Cherwell/Cherwell.py]: 👍

File: Scripts/CherwellGetIncident/CherwellGetIncident.py [hunk 14]
```diff
@@ -0,0 +1,90 @@
...
+    'PublicId',
+    'Description',
+    'Priority',
+    'CustomerDisplayName',
+    'OwnedBy',
+    'Service',
+    'CreatedDateTime',
+    'TotalTasks'
+]
+
```
Comment (yuvalbenshalom) [inline Scripts/CherwellGetIncident/CherwellGetIncident.py]: I think it is a good idea to put the output path prefix here in a separate var so the user will be abele to change it.
`output_path = "Cherwell.BusinessObjects"`

File: Scripts/CherwellIncidentOwnTask/CherwellIncidentOwnTask.py [hunk 15]
```diff
@@ -0,0 +1,48 @@
...
+
+# ####################################################################################
+# ############################## CONFIGURATION PART ##################################
+# ####################################################################################
+
+"""
+`RELATIONSHIP_ID` is the id of the relationship you wish to base your link on. In this case we wanted to create script
+for incidents to own tasks, thus, we used the "Incident Owns Task" relationship id.
+"""
+RELATIONSHIP_ID = '9369187528b417b4a17aaa4646b7f7a78b3c821be9'
```
Comment (yuvalbenshalom) [inline Scripts/CherwellIncidentOwnTask/CherwellIncidentOwnTask.py]: put this as an exception for guard-rails

File: Scripts/CherwellIncidentUnlinkTask/CherwellIncidentUnlinkTask.py [hunk 16]
```diff
@@ -0,0 +1,48 @@
...
+
+# ####################################################################################
+# ############################## CONFIGURATION PART ##################################
+# ####################################################################################
+
+"""
+`RELATIONSHIP_ID` is the id of the relationship you wish to base your link on. In this case we wanted to create script
+for incidents to own tasks, thus, we used the "Incident Owns Task" relationship id.
+"""
+RELATIONSHIP_ID = '9369187528b417b4a17aaa4646b7f7a78b3c821be9'
```
Comment (yuvalbenshalom) [inline Scripts/CherwellIncidentUnlinkTask/CherwellIncidentUnlinkTask.py]: put this as an exception for guard-rails

File: Scripts/CherwellQueryIncidents/CherwellQueryIncidents.py [hunk 17]
```diff
@@ -0,0 +1,87 @@
...
+    'PublicId',
+    'Description',
+    'Priority',
+    'CustomerDisplayName',
+    'OwnedBy',
+    'Service',
+    'CreatedDateTime',
+    'TotalTasks'
+]
+
```
Comment (yuvalbenshalom) [inline Scripts/CherwellQueryIncidents/CherwellQueryIncidents.py]: I think it is a good idea to put the output path prefix here in a separate var so the user will be abele to change it.
output_path = "Cherwell.BusinessObjects"

File: Scripts/CherwellGetIncident/CherwellGetIncident.py (hunk 14 above)
Comment (ShahafBenYakir) [reply Scripts/CherwellGetIncident/CherwellGetIncident.py]: - [x] Check

File: Scripts/CherwellQueryIncidents/CherwellQueryIncidents.py (hunk 17 above)
Comment (ShahafBenYakir) [reply Scripts/CherwellQueryIncidents/CherwellQueryIncidents.py]: - [x] Check

File: Integrations/Cherwell/Cherwell.yml [hunk 18]
```diff
@@ -0,0 +1,526 @@
...
+      required: true
+      secret: false
+    - default: false
+      description: Public id or record id
+      isArray: false
+      name: id_value
+      required: true
+      secret: false
+    deprecated: false
+    description: Doenload imported attachements from given business object.
```
[the remark this answers is not in the document]
Comment (ShahafBenYakir) [reply Integrations/Cherwell/Cherwell.yml]: - [x] Check

File: Integrations/Cherwell/Cherwell.yml [hunk 19]
```diff
@@ -0,0 +1,526 @@
+category: IT Services
+commonfields:
+  id: Cherwell
+  version: -1
+configuration:
+- display: url
```
[the remark this answers is not in the document]
Comment (ShahafBenYakir) [reply Integrations/Cherwell/Cherwell.yml]: - [ ] Check

File: Integrations/Cherwell/Cherwell.py (hunk 12 above)
Comment (ShahafBenYakir) [reply Integrations/Cherwell/Cherwell.py]: - [x] Check

File: Integrations/Cherwell/Cherwell.py (hunk 10 above)
Comment (ShahafBenYakir) [reply Integrations/Cherwell/Cherwell.py]: - [x]

File: Integrations/Cherwell/Cherwell.py (hunk 1 above)
Comment (ShahafBenYakir) [reply Integrations/Cherwell/Cherwell.py]: - [x] Check

File: Integrations/Cherwell/Cherwell.py (hunk 2 above)
Comment (ShahafBenYakir) [reply Integrations/Cherwell/Cherwell.py]: - [x] Check

File: Integrations/Cherwell/Cherwell.py (hunk 3 above)
Comment (ShahafBenYakir) [reply Integrations/Cherwell/Cherwell.py]: - [x] Check

File: Integrations/Cherwell/Cherwell.py (hunk 4 above)
Comment (ShahafBenYakir) [reply Integrations/Cherwell/Cherwell.py]: - [x] Check

File: Integrations/Cherwell/Cherwell.py (hunk 5 above)
Comment (ShahafBenYakir) [reply Integrations/Cherwell/Cherwell.py]: - [x] Check

File: Integrations/Cherwell/Cherwell.py (hunk 6 above)
Comment (ShahafBenYakir) [reply Integrations/Cherwell/Cherwell.py]: - [x] Check

File: Integrations/Cherwell/Cherwell.py (hunk 7 above)
Comment (ShahafBenYakir) [reply Integrations/Cherwell/Cherwell.py]: - [x] Check

File: Integrations/Cherwell/Cherwell.py (hunk 8 above)
Comment (ShahafBenYakir) [reply Integrations/Cherwell/Cherwell.py]: - [x] Check

File: Integrations/Cherwell/Cherwell.py (hunk 9 above)
Comment (ShahafBenYakir) [reply Integrations/Cherwell/Cherwell.py]: - [x] Check

File: Integrations/Cherwell/Cherwell.py (hunk 11 above)
Comment (ShahafBenYakir) [reply Integrations/Cherwell/Cherwell.py]: - [x] Check

File: Integrations/Cherwell/Cherwell.py [hunk 20]
```diff
@@ -0,0 +1,992 @@
...
+
+def get_new_access_token():
+    response = request_new_access_token(True)
+    if not response.status_code == HTTP_CODES['success']:
+        response = request_new_access_token(False)
+    res_json = parse_response(response,
+                              "Could not get token. Check your credentials (user/password/client id) and try again")
+    demisto.setIntegrationContext({
+        'refresh_token': res_json.get('refresh_token'),
+        'token_expiration_time': int(date_to_timestamp(res_json.get('.expires'), '%a, %d %b %Y %H:%M:%S GMT')),
```
[the remark this answers is not in the document]
Comment (ShahafBenYakir) [reply Integrations/Cherwell/Cherwell.py]: - [x] Check

File: Integrations/Cherwell/Cherwell.py [hunk 21]
```diff
@@ -0,0 +1,992 @@
...
+        return_error('Error connecting to server. Check your URL/Proxy/Certificate settings')
+    return response
+
+
+def request_new_access_token(using_refresh):
+    url = BASE_URL + "token"
+    refresh_token = demisto.getIntegrationContext().get('refresh_token')
+
+    payload = f'client_id={CLIENT_ID}'
+    payload = payload + f'&grant_type=refresh_token&refresh_token={refresh_token}' if using_refresh \
```
[the remark this answers is not in the document]
Comment (ShahafBenYakir) [reply Integrations/Cherwell/Cherwell.py]: = [x] Check

File: Integrations/Cherwell/Cherwell.py [hunk 22]
```diff
@@ -0,0 +1,992 @@
...
+    except requests.exceptions.ConnectionError as e:
+        return_error('Error connecting to server. Check your URL/Proxy/Certificate settings')
+    return response
+
+
+def request_new_access_token(using_refresh):
+    url = BASE_URL + "token"
+    refresh_token = demisto.getIntegrationContext().get('refresh_token')
+
+    payload = f'client_id={CLIENT_ID}'
```
[the remark this answers is not in the document]
Comment (ShahafBenYakir) [reply Integrations/Cherwell/Cherwell.py]: - [x] Check

File: Scripts/CherwellIncidentOwnTask/CherwellIncidentOwnTask.py (hunk 15 above)
Comment (ShahafBenYakir) [reply Scripts/CherwellIncidentOwnTask/CherwellIncidentOwnTask.py]: - [x] Check

File: Scripts/CherwellIncidentUnlinkTask/CherwellIncidentUnlinkTask.py (hunk 16 above)
Comment (ShahafBenYakir) [reply Scripts/CherwellIncidentUnlinkTask/CherwellIncidentUnlinkTask.py]: - [x] Check

File: Integrations/Cherwell/Cherwell.py [hunk 23]
```diff
@@ -0,0 +1,992 @@
...
+    except requests.exceptions.HTTPError:
+        try:
+            res_json = response.json()
+            err_msg = res_json.get('errorMessage') or res_json.get('error_description') or res_json.get('Message')
+        except Exception:
+            err_msg = response.content.decode('utf-8')
+        return_error(error_operation + ": " + err_msg)
+    except Exception:
+        try:
+            return response.content  # check if needed
```
[the remark this answers is not in the document]
Comment (ShahafBenYakir) [reply Integrations/Cherwell/Cherwell.py]: the first `res.json()`

File: Integrations/Cherwell/Cherwell.py [hunk 24]
```diff
@@ -0,0 +1,992 @@
...
+
+
+''' HELPER FUNCTIONS '''
+
+
+def parse_response(response, error_operation):
+    try:
+        response.raise_for_status()
+        if not response.content:
+            return
```
[the remark this answers is not in the document]
Comment (ShahafBenYakir) [reply Integrations/Cherwell/Cherwell.py]: yes some responses are empty

File: Integrations/Cherwell/Cherwell.py [hunk 25]
```diff
@@ -0,0 +1,992 @@
...
+
+QUERY_OPERATORS = ['eq', 'gt', 'lt', 'contains', 'startwith']
+
+#######################################################################################################################
+
+
+''' HELPER FUNCTIONS '''
+
+
+def parse_response(response, error_operation):
```
[the remark this answers is not in the document]
Comment (ShahafBenYakir) [reply Integrations/Cherwell/Cherwell.py]: I simplified it a little bit. Unfortunately, Cherwell API returns their responses in several ways so I wrote one function that will parse all responses. it makes it easier to manage it like this.

**Pull request merged**